### ISSUE TITLE
Vertical alignment in format expression

### ISSUE THREAD

**Issue by stanislawpuda-tomtom** [maplibre/maplibre-style-spec#832]
# Design Proposal: Add vertical alignment to format expression

## Motivation

[Format expression](https://maplibre.org/maplibre-style-spec/expressions/#format) can facilitate multiple text and image elements of different sizes. At the moment all of them are aligned to the font baseline. It is not possible to change the vertical alignment. Having a possibility to change the alignment, eg. center, would allow us to create rich labels.

If following design would be accepted I would like to make contributions to this repo and MapLibre GL JS with the implementation.

## Proposed Change

We would like to introduce `vertical-align` property to style override in `format` expression. `vertical-align` property can take three options:
- `bottom` *(default)*
- `center`
- `top`

#### Center

```JavaScript
"text-field": [
  "format",
  ["image", "image"], 
  { "vertical-align": "center" },
  " ",
  {},
  "BIG yg",
  { "font-scale": 1.2, "vertical-align": "center" },
  " ",
  {},
  "SMALL",
  { "font-scale": 0.8, "vertical-align": "center" }
],
```

![big_image_center](https://github.com/user-attachments/assets/4dd7a13f-62cc-4d0d-9d39-c73d67bc7009)
![small_image_center](https://github.com/user-attachments/assets/cd54deb5-f27f-4bd3-b82e-6eb8efd5a474)

#### Top

```JavaScript
"text-field": [
  "format",
  ["image", "image"], 
  { "vertical-align": "top" },
  " ",
  {},
  "BIG yg",
  { "font-scale": 1.2, "vertical-align": "top" },
  " ",
  {},
  "SMALL",
  { "font-scale": 0.8, "vertical-align": "top" }
],
```

![big_image_top](https://github.com/user-attachments/assets/7fad5463-f61e-4f23-9410-7083302c21fc)
![small_image_top](https://github.com/user-attachments/assets/ecebf3d4-a83b-4c11-b392-c386dc7eb567)

#### Bottom (*default*, current behaviour)

```JavaScript
"text-field": [
  "format",
  ["image", "image"], 
  {},
  " ",
  {},
  "BIG yg",
  { "font-scale": 1.2 },
  " ",
  {},
  "SMALL",
  { "font-scale": 0.8 }
],
```

![big_image_baseline](https://github.com/user-attachments/assets/435f8267-9348-4d07-957a-48bd71ef5f4e)
![small_image_baseline](https://github.com/user-attachments/assets/3d2b4481-e4bf-4b4c-8ba8-ed2aef2465f4)

#### Mixed

```JavaScript
"text-field": [
  "format",
  ["image", "image"], 
  {},
  " ",
  {},
  "BIG yg",
  { "font-scale": 1.2 },
  " ",
  {},
  "SMALL",
  { "font-scale": 0.8, "vertical-align": "center" }
],
```

![big_image_mixed](https://github.com/user-attachments/assets/e4ab7f4c-f911-49a1-b406-c776a331abc5)

## API Modifications

Add new property to style override object of `format` expression:

`vertical-align`:
- type: `enum`
- default: `bottom`
- possible values: `bottom`, `center`, `top`
- effect: change vertical alignment of `format` expression element

#### Behaviour

- `bottom` *default*: text [baseline](https://en.wikipedia.org/wiki/Baseline_(typography)) and image bottom are in line (current behaviour)
- `center`: image center and text center (line between baseline and top) are in line with center of the highest element in the row
- `top`: image top and text top (line determined by the highest character in the font) are in line

It is possible to set `vertical-align` for each element in `format` expression separately.

## Migration Plan and Compatibility

There is no breaking change. Rendering of existing styles is unchanged. Only if style has new property `vertical-align` with non-default value, there is a difference in label rendering.

**Comment by HarelM** [comment]
Thanks for taking the time to open the proposal! 
What are the current alternatives available? 
My concern is the this is a bit two simplistic. What about if I want to move a single word two pixels upwards? What about padding between words?
This might get very complex...

**Comment by stanislawpuda-tomtom** [comment]
Unfortunately we couldn't find any alternative. Format expression does not offer any control over Y position of elements, they are always aligned to the font baseline.

I actually value simplicity in this solution, however I understand your concern. We did consider one solution based on relative offset, where numeric value tells how much of the element should be moved - example:
- 0 - bottom of the image in line with the font baseline
- 0.5 - middle of the image in line with the font baseline
- 1 - top of the image in line with the font baseline 

However this way is very difficult to actually align elements, if images vary in size. This is why we choose simple option.

We could try to combine both approaches, and see if alignment can be combined with offset.

What is your opinion on that?

**Comment by stanislawpuda-tomtom** [comment]
One other option that comes to my mind is to make something like numerical alignment, eg:

0 - element aligned to the baseline / bottom of the highest element
0.5 - element aligned to the center of the highest element
1 - element aligned to the top of the highest element.

This could work similar to `baseline`, `center` and `top`, while giving more control. However, I'm not sure if more declarative way isn't more intuitive.

**Comment by stanislawpuda-tomtom** [comment]
I understand your concern. I believe what you've mention can be achieved by combining together additional property. 

If we want to move single word a few pixels, we could add `vertical-offset`. Examples:
- element aligned to the top and shifted two pixels down: `{ "vertical-align": "top", "vertical-offset": "-2px" }`
- element aligned to the center and shifted two pixels up: `{ "vertical-align": "center", "vertical-offset": "2px" }`
"offset" property maybe could take also relative values, but it should be well thought out to make it intuitive.

For the padding between words the easiest solution is to use spaces. I'm not sure if this requires another solution.

For the alternatives, we did think of numeric based approach, where value moves element up and down. There are two possibilities:
- where numeric value stands for relation between baseline of the row and element vertical position. Eg. 0 - baseline of the element = baseline of the row, 0.5 - element shifted 50% of its height up, -0.5 or `center`, middle of the element in line with row baseline. However, this approach is not working when elements (especially images) can have various height.
- similar to `vertical-align` but with numbers: 0 for baseline, 0.5 for middle and 1 for top - but enum variant is in my opinion easier to understand

**Comment by zbigniewmatysek-tomtom** [comment]
I propose to rename this property, because it suggests it works like https://developer.mozilla.org/en-US/docs/Web/CSS/vertical-align, but in reality is similar to the https://developer.mozilla.org/en-US/docs/Web/CSS/align-self.

I belive showing the analogy to CSS properties can be helpful. It is easier to explain and adopt trough using existing terminology.

Regarding the @HarelM concern on the extendibility of this proposal. I think it's pretty extensible. One solution is having relative offset property that @stanislawpuda-tomtom explained or a different set of properties altogether that may or may not be combained with the vertical-align.

**Comment by HarelM** [comment]
I would advise to bring this up in the monthly meeting to discuss this.

**Comment by stanislawpuda-tomtom** [comment]
Will do!

Regarding the name, I went through few possible options:
- `row-align`,
- `align-vertical`,
- `vertical-placement`,
- `y-align`?

But to be honest, `vertical-align` still sounds for me like the best option. I'm not sure if the connotations with CSS property is so significant.

**Comment by stanislawpuda-tomtom** [comment]
@HarelM

Done, I've changed the PR.

**Comment by HarelM** [comment]
@zbigniewmatysek-tomtom any feedback on the above?

**Comment by zbigniewmatysek-tomtom** [comment]
@HarelM 
Baseline aligns bottom of line elements to the baseline of the text, this is not the same as the bottom alignment which aligns elements to the bottom of the tallest element.

If we decide to extend spec to support bottom alignment, it will be cumbersome to name it.

I like the "Middle", since it reflects that alignment is in a single axis.

How about `Top/Middle/Baseline`?

And in the future (if someone finds it useful): `Top/Middle/Baseline/Bottom`

**Comment by HarelM** [comment]
I'm not sure baseline is intuitive, it's not intuitive to me at least...

**Comment by zbigniewmatysek-tomtom** [comment]
LGTM. 

I also don't see an use case for the bottom, but I wanted to raise this just in case. We can go with your proposal.

**Comment by louwers** [comment]
> I tend to agree that when I read the CSS of self-align I have no understanding of what self means in that aspect...

It's the piece of text or image that the format override applies to.

By the way, there is a CSS `vertical-align`, why not use the same terms as that? https://developer.mozilla.org/en-US/docs/Web/CSS/vertical-align

![Image](https://github.com/user-attachments/assets/ad08d2b8-e141-4012-b197-7f37153ba135)

**Comment by zbigniewmatysek-tomtom** [comment]
@louwers 

https://github.com/maplibre/maplibre-style-spec/issues/832#issuecomment-2376619999

The behavior of css `vertical-align` is very different from the behavior we want to achieve here.

**Comment by louwers** [comment]
I think I understand what you mean. `vertical-align: middle` does the following:

> Aligns the middle of the element with the baseline plus half the x-height of the parent.

![Image](https://github.com/user-attachments/assets/937baaa6-0b5e-446f-9df4-a336c2fb02c4)

Here pink is `middle`, orange is `top` and green is `baseline`. The SMALL is not in the middle of the entire line.

In fact, `vertical-align` does not have a line-relative value for the center/middle:

![Image](https://github.com/user-attachments/assets/33f14590-e1d7-4eb4-9ec3-278ff9d592cb)

It does for table cells though:

![Image](https://github.com/user-attachments/assets/84fffac4-c7a9-467a-b8ad-461107940247)

https://developer.mozilla.org/en-US/docs/Web/CSS/vertical-align

**Comment by stanislawpuda-tomtom** [comment]
@louwers 
> I assume the tallest image of the format expression also participates in determining the top line?

This is correct - the highest section of the line, can be either image or text
